fault-remediation/pkg/reconciler/reconciler.go

```diff
@@ -64,96 +60,79 @@ type HealthEventData struct {
+	annotationManager NodeAnnotationManagerInterface
+	dryRun            bool
 }
 
-func (r *Reconciler) Start(ctx context.Context) error {
-	// Create datastore instance
-	ds, err := datastore.NewDataStore(ctx, r.Config.DataStoreConfig)
-	if err != nil {
-		return fmt.Errorf("error initializing datastore: %w", err)
-	}
-
-	defer func() {
-		if err := ds.Close(ctx); err != nil {
-			slog.Error("failed to close datastore", "error", err)
-		}
-	}()
-
-	// Create watcher using the factory pattern
-	watcherConfig := watcher.WatcherConfig{
-		Pipeline:       r.Config.Pipeline,
-		CollectionName: "HealthEvents",
-	}
-
-	watcherInstance, err := watcher.CreateChangeStreamWatcher(ctx, ds, watcherConfig)
-	if err != nil {
-		return fmt.Errorf("error initializing change stream watcher: %w", err)
-	}
-
-	defer func() {
-		if err := watcherInstance.Close(ctx); err != nil {
-			slog.Error("failed to close watcher", "error", err)
-		}
-	}()
-
-	// Get the HealthEventStore for document operations
-	healthEventStore := ds.HealthEventStore()
-
-	watcherInstance.Start(ctx)
-	slog.Info("Listening for events on the channel...")
-
-	for event := range watcherInstance.Events() {
-		slog.Info("Event received", "event", event)
-		r.processEvent(ctx, event, watcherInstance, healthEventStore)
+// TODO: reduce args to this function call by passing Reconciler Params
+func NewFaultRemediationReconciler(
+	ds datastore.DataStore,
+	watcher datastore.ChangeStreamWatcher,
+	healthEventStore datastore.HealthEventStore,
+	config ReconcilerConfig,
+	dryRun bool,
+) FaultRemediationReconciler {
+	return FaultRemediationReconciler{
+		ds:                ds,
+		watcher:           watcher,
+		healthEventStore:  healthEventStore,
+		config:            config,
+		annotationManager: config.RemediationClient.GetAnnotationManager(),
+		dryRun:            dryRun,
 	}
-
-	return nil
 }
 
-// processEvent handles a single event from the watcher
-func (r *Reconciler) processEvent(ctx context.Context, eventWithToken datastore.EventWithToken,
-	watcherInstance datastore.ChangeStreamWatcher, healthEventStore datastore.HealthEventStore) {
+// TODO: its possible to have controller-runtime generate all necessary rbac from code annotations here
+func (r *FaultRemediationReconciler) Reconcile(
+	ctx context.Context,
+	event *datastore.EventWithToken,
+) (ctrl.Result, error) {
+	//TODO: may not need these metrics anymore if we have some defaults from ctrl runtime
 	start := time.Now()
 
+	slog.Info("Reconciling Event")
+
 	defer func() {
 		eventHandlingDuration.Observe(time.Since(start).Seconds())
 	}()
 
 	totalEventsReceived.Inc()
 
-	healthEventWithStatus, err := r.parseHealthEvent(eventWithToken, watcherInstance)
+	healthEventWithStatus, err := r.parseHealthEvent(*event, r.watcher)
 	if err != nil {
-		return
+		//TODO: error handling?
+		return ctrl.Result{}, nil
 	}
 
 	// Safety checks for nil pointers
 	if healthEventWithStatus.HealthEvent == nil {
+		//TODO: error handling?
```

Comment:
This was a defensive check; this ideally should not happen. So, in this case too we want to drop the event with a warning log